### PR TITLE
Comment Analysis By Category

### PR DESCRIPTION
**Pull request by olu144**
Users can now select a category they would like to analyze the comments of using radio buttons. The functionality of inputting a username of channel id is preserved and there is a radio button to specify this type of input. 
Screenshot 2020-07-13 at 7 51 42 PM - Display 2

### CODE REVIEW

File: capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java [hunk 1]
```diff
@@ -43,7 +43,13 @@
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String ChannelId = request.getParameter("channelId");
-    String url = URL + ChannelId + "&maxResults=" + NUM_RESULTS + "&key=" + KEY;
+    String videoID = request.getParameter("videoID");
```
Comment (danieldortega77) [inline capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java]: nit: Let's keep a consistent naming style between `ChannelId` and `videoID` - I would recommend `channelId` and `videoId`

File: capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  /** Makes a GET request. */
+  private String get(String url) throws IOException {
+    Request request = new Request.Builder()
+      .url(url)
+      .build();
+    try (Response response = client.newCall(request).execute()) {
+      return response.body().string();
+    }
+  }
+}
```
Comment (danieldortega77) [inline capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java]: nit: This should have a terminating newline

File: capstone/src/main/webapp/youtubeScript.js [hunk 3]
```diff
@@ -174,3 +198,75 @@ function isLetter(character) {
     return false;
   }
 }
+
+async function getTrending(categoryId) {
+  trendingResponse = await fetch('/trending_servlet?videoCategoryId=' + categoryId,)
+  trendingResponseJson = await trendingResponse.json();
```
Comment (danieldortega77) [inline capstone/src/main/webapp/youtubeScript.js]: I think we should declare these with `const`, unless it produces an error I'm not aware of.

File: capstone/src/main/webapp/youtubeScript.js [hunk 4]
```diff
@@ -174,3 +198,75 @@ function isLetter(character) {
...
+  const trendingVideoIds = [];
+  for (const item in trendingResponseJson.items) {
+    const videoId = trendingResponseJson.items[item].id;
+    trendingVideoIds.push(videoId);
+  }
+  const commentsList = []
+  for (const id in trendingVideoIds) {
+    videoCommentList = await fetch('/youtube_servlet?videoID=' + trendingVideoIds[id],)
+    videoCommentListJson = await videoCommentList.json();
```
Comment (danieldortega77) [inline capstone/src/main/webapp/youtubeScript.js]: Same here - we should declare these with `const`

File: capstone/src/main/webapp/youtubeScript.js [hunk 5]
```diff
@@ -21,14 +21,34 @@ const ATTRIBUTES_BY_LANGUAGE = {
...
+  'How-to&Style' : 26,
+  'Music': 10,
+  'Pets&Animals': 15,
+  'Science&Technology': 28,
+  'Sports' : 17,
+};
+
 /** Calls youtube servlet and passes output to perspctive */
 async function callYoutube() {
+  document.getElementById('search-type').innerHTML = ""
```
Comment (danieldortega77) [inline capstone/src/main/webapp/youtubeScript.js]: nit: missing a terminating ";"

File: capstone/src/main/webapp/youtubeScript.js [hunk 6]
```diff
@@ -45,15 +66,16 @@ async function callYoutube() {
       inputCommentsToPerspective([]);
       return;
     }
+    document.getElementById('search-type').innerHTML ="Username Search";
```
Comment (danieldortega77) [inline capstone/src/main/webapp/youtubeScript.js]: nit for lines 46, 60, and 69: Let's add a space after the `=`

File: capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+@WebServlet("/trending_servlet")
+public class YoutubeTrendingServlet extends HttpServlet {
+  private static final String URL = "https://www.googleapis.com/youtube/v3/videos?part=snippet%2CcontentDetails%2Cstatistics&chart=mostPopular";
+  private static final String KEY = "API_KEY";
+  private static final String NUM_RESULTS = "2";
+  private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
+  OkHttpClient client = new OkHttpClient();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (danieldortega77) [inline capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java:42]: I think, depending on structure, that we could make this a POST request (although it would require some changing youtubeScript.js) so that users cannot use this servlet as a free means of accessing the YouTube API with our key. The same suggestion applies to YoutubeServlet.java.

File: capstone/src/main/webapp/youtubeScript.js (hunk 6 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js]: addressed in "spacing and address Daniel's nits"

File: capstone/src/main/webapp/youtubeScript.js (hunk 5 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js]: addressed in "spacing and address Daniel's nits"

File: capstone/src/main/webapp/youtubeScript.js (hunk 4 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js]: addressed in "spacing and address Daniel's nits"

File: capstone/src/main/webapp/youtubeScript.js (hunk 3 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js]: addressed in "spacing and address Daniel's nits"

File: capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java (hunk 2 above)
Comment (olu144) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java]: addressed in "spacing and address Daniel's nits"

File: capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java (hunk 1 above)
Comment (olu144) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java]: addressed in "spacing and address Daniel's nits"

File: capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java (hunk 7 above)
Comment (olu144) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java:42]: I will look into making that change.

File: capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java [hunk 8]
```diff
@@ -43,7 +43,13 @@
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String ChannelId = request.getParameter("channelId");
-    String url = URL + ChannelId + "&maxResults=" + NUM_RESULTS + "&key=" + KEY;
+    String videoId = request.getParameter("videoId");
+    String url;
+    if (ChannelId!=null) {
+      url = URL + "&allThreadsRelatedToChannelId=" + ChannelId + "&maxResults=" + NUM_RESULTS + "&key=" + KEY;
```
Comment (chrismohrii) [inline capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java]: nit: Could we change the variable names of `url` and/or `URL` so it's easier to see what the difference between them is?

File: capstone/src/main/webapp/youtubeScript.js [hunk 9]
```diff
@@ -21,22 +21,43 @@ const ATTRIBUTES_BY_LANGUAGE = {
   'pt': ['TOXICITY', 'SEVERE_TOXICITY', 'IDENTITY_ATTACK', 'INSULT', 'PROFANITY', 'THREAT']
 };
 
+const YOUTUBE_CATEGORIES = {
```
Comment (chrismohrii) [inline capstone/src/main/webapp/youtubeScript.js:25]: nit: Let's put a comment to say what these integer values represent

File: capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java [hunk 10]
```diff
@@ -43,7 +43,13 @@
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String ChannelId = request.getParameter("channelId");
-    String url = URL + ChannelId + "&maxResults=" + NUM_RESULTS + "&key=" + KEY;
+    String videoId = request.getParameter("videoId");
+    String url;
+    if (ChannelId!=null) {
```
Comment (chrismohrii) [inline capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java]: nit: Let's put spaces before and after `!=`

File: capstone/src/main/webapp/youtubeScript.js [hunk 11]
```diff
@@ -64,24 +86,26 @@ async function inputCommentsToPerspective(comments) {
...
-    attributeScores.push(perspectiveScore);
-  }
   const attributeTotals = new Map();
-  for (var i = 0; i < requestedAttributes.length; i++) {
-    for (var j = 0; j < attributeScores.length; j++) {
-      if (attributeTotals.has(requestedAttributes[i])) {
-        attributeTotals.set(requestedAttributes[i], attributeTotals.get(requestedAttributes[i]) + attributeScores[j].attributeScores[requestedAttributes[i]].summaryScore.value);
-      } else {
-        attributeTotals.set(requestedAttributes[i], attributeScores[j].attributeScores[requestedAttributes[i]].summaryScore.value);
+  for (const comments in commentsList) {
```
Comment (chrismohrii) [inline capstone/src/main/webapp/youtubeScript.js:90]: Could we put these average calculations into a separate helper method to make this one a little more readable?

File: capstone/src/main/webapp/youtubeScript.js [hunk 12]
```diff
@@ -174,3 +198,75 @@ function isLetter(character) {
...
+  radiobox.name = 'same';
+  radiobox.checked  = true;
+  const label = document.createElement('label');
+  label.htmlFor = 'manualInput';
+  const description = document.createTextNode('ID/Username');
+  label.appendChild(description);
+  radiobox.onclick = function() {
+    disableTextInput(this);   
+  }
+  const container = document.getElementById('category-container');
```
Comment (chrismohrii) [inline capstone/src/main/webapp/youtubeScript.js]: nit: Let's rename `container` to something a little more descriptive like `categoryContainer`

Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js]: addressed in "address Chris's nits"

File: capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java (hunk 10 above)
Comment (olu144) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java]: addressed in "address Chris's nits"

File: capstone/src/main/webapp/youtubeScript.js (hunk 9 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js:25]: addressed in "address Chris's nits"

File: capstone/src/main/webapp/youtubeScript.js (hunk 11 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js:90]: addressed in "separate into classes and utilize promise"

File: capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java (hunk 8 above)
Comment (danieldortega77) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java]: +1

File: capstone/src/main/webapp/youtubeScript.js [hunk 13]
```diff
@@ -21,39 +21,61 @@ const ATTRIBUTES_BY_LANGUAGE = {
...
-  var response;
+  /** Checks if input is a category, if so directs input to be handled by get trending*/
+  if (YOUTUBE_CATEGORIES[channelId] != undefined) {
+    document.getElementById('search-type').innerHTML = "Category Search";
+    getTrending(YOUTUBE_CATEGORIES[channelId]);
+    return;
+  }
   /** Checks if input follows channel ID format, if not attempts to convert it to channel ID*/
+  var response;
+  var responseJson;
```
Comment (danieldortega77) [inline capstone/src/main/webapp/youtubeScript.js]: nit: The Google Style Guide suggests we always use `let` instead of `var`, so I would change these, as well as the declarations in lines 108 and 109.

File: capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java (hunk 8 above)
Comment (olu144) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java]: addressed in "improve variable names"

File: capstone/src/main/webapp/youtubeScript.js (hunk 13 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js]: addressed in "change var to let"

**Review by danieldortega77 — APPROVED** [review]

File: capstone/src/main/webapp/youtubeScript.js [hunk 14]
```diff
@@ -162,15 +202,83 @@ function showAvailableAttributes() {
...
 
 /** Checks if a character is a letter */
 function isLetter(character) {
-  if ((character.charCodeAt() >= 65 && character.charCodeAt() <= 90) || (character.charCodeAt() >= 97 && character.charCodeAt() <= 122)) {
-    return true;
-  } else {
-    return false;
+  return (character.charCodeAt() >= 65 && character.charCodeAt() <= 90) || (character.charCodeAt() >= 97 && character.charCodeAt() <= 122); 
+}
+
```
Comment (danieldortega77) [inline capstone/src/main/webapp/youtubeScript.js:215]: Let's add some comments explaining what this function and the ones below it do

**Review by danieldortega77 — APPROVED** [review]

File: capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java [hunk 15]
```diff
@@ -43,8 +43,14 @@
...
     String ChannelId = request.getParameter("channelId");
-    String url = URL + ChannelId + "&maxResults=" + NUM_RESULTS + "&key=" + KEY;
-    String output = get(url);
+    String videoId = request.getParameter("videoId");
+    String completeUrl;
+    if (ChannelId != null) {
+      completeUrl = BASE_URL + "&allThreadsRelatedToChannelId=" + ChannelId + "&maxResults=" + NUM_RESULTS + "&key=" + KEY;
+    } else {
+      completeUrl = BASE_URL + "&videoId=" + videoId + "&maxResults=" + NUM_RESULTS + "&key=" + KEY;
+    }
```
Comment (chrismohrii) [inline capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java]: nit: This could be simplified a bit with the ternary operator

https://www.geeksforgeeks.org/java-ternary-operator-with-examples/

File: capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java (hunk 7 above)
Comment (chrismohrii) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java:42]: +1

File: capstone/src/main/webapp/youtubeScript.js [hunk 16]
```diff
@@ -162,15 +202,83 @@ function showAvailableAttributes() {
     label.appendChild(document.createTextNode(attribute));
     avaiableAttributesElement.appendChild(checkbox);
     avaiableAttributesElement.appendChild(label);
-    avaiableAttributesElement.appendChild (document.createTextNode (" "));
+    avaiableAttributesElement.appendChild(document.createTextNode (" "));
```
Comment (chrismohrii) [inline capstone/src/main/webapp/youtubeScript.js:207]: nit: Let's delete the extra space after `createTextNode` and on line 264 & 282

File: capstone/src/main/webapp/youtubeScript.js [hunk 17]
```diff
@@ -162,15 +202,83 @@ function showAvailableAttributes() {
...
+    document.getElementById('channelIdForAnalysis').disabled = true;   
+  }
+}
+
+function disableTextInput(button) {
+  if (button.checked) { 
+    document.getElementById('channelIdForAnalysis').value = "";
+    document.getElementById('channelIdForAnalysis').disabled = false;
+  }
+}
```
Comment (chrismohrii) [inline capstone/src/main/webapp/youtubeScript.js]: nit: These two methods could be condensed into one and you could pass in a `value` and `isDisabled` parameter to avoid code repetition

File: capstone/src/main/webapp/youtubeScript.js [hunk 18]
```diff
@@ -162,15 +202,83 @@ function showAvailableAttributes() {
...
+
+/** Creates radio buttons to allow teh user to select between various categories*/
```
Comment (chrismohrii) [inline capstone/src/main/webapp/youtubeScript.js]: nit: `teh` -> `the`

File: capstone/src/main/webapp/youtubeScript.js (hunk 14 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js:215]: addressed in "address chris and daniel nits"

File: capstone/src/main/webapp/youtubeScript.js (hunk 18 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js]: addressed in "address chris and daniel nits"

File: capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java (hunk 15 above)
Comment (olu144) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeServlet.java]: addressed in "address chris and daniel nits"

File: capstone/src/main/webapp/youtubeScript.js (hunk 16 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js:207]: addressed in "remove extra space

File: capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java (hunk 7 above)
Comment (olu144) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java:42]: After some research, I've discovered that GET and POST requests are not interchangeable. The type of request is dependent on the API and the wishes of the creators. Ill provide updates if I find information that suggests otherwise.

File: capstone/src/main/webapp/youtubeScript.js (hunk 17 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js]: addressed in "consolidate text input toggle function"
instead of passing in a value I decided to pass in the button as I felt that would make the function easier to use and build upon.

File: capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java (hunk 7 above)
Comment (chrismohrii) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java:42]: I think we would just want change this GET method to POST in the servlet, Youtube call stays the same

File: capstone/src/main/webapp/youtubeScript.js [hunk 19]
```diff
@@ -65,20 +87,24 @@ async function inputCommentsToPerspective(comments) {
...
+      const perspectiveScore = await callPerspective(commentsList[comments].items[item].snippet.topLevelComment.snippet.textOriginal, langElement.value, requestedAttributes);
+      attributeScoresPromises.push(perspectiveScore);
+    }
   }
   await Promise.all(attributeScoresPromises).then(resolvedAttributeScores => {
-    const attributeAverages = getAttributeAverages(resolvedAttributeScores, comments.items.length);
+    const attributeTotals = getAttributeTotals(resolvedAttributeScores);
+    const attributeAverages = getAttributeAverages(attributeTotals, commentsList);
     loadChartsApi(attributeAverages);
+    perspectiveToxicityScale(attributeAverages);
```
Comment (chrismohrii) [inline capstone/src/main/webapp/youtubeScript.js]: When I run this code, I'm getting the error that this method is undefined, and I don't see its implementation in this file. Is it maybe in another branch?

Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js]: I'm unsure how that got there. That function is implemented in branch 21. I removed the problematic line in "remove call to perspectiveToxicityScale"
Also, I will transition from GET requests to POST request in a new PR as the change requires many files to be changed.

**Review by chrismohrii — APPROVED** [review]

File: capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java (hunk 7 above)
Comment (dborkan) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeTrendingServlet.java:42]: This conversation is good and I'm glad you are all thinking about security!  That said I think:
* The YouTube API (i.e. the request to `https://www.googleapis.com/youtube/v3/...`) seems to require a GET request, so I'm not sure we can change that.  It is possible that our servlet could implement a `doPost` instead of `doGet` regardless of what the call to youtube is.  `doPost` might make things more consistent with our other calls to the servlet, but I don't think it's really required.
* GET is typically used when just fetching some data, rather than doing an action that might change state on the server, so I do think it may make sense here (it's a little unintuitive that perspective uses POST TBH).  POST makes more sense for actions that add or update something on the server (like adding comments to a message board)
* Regarding getting free access to the YouTube API, the usual solution to this is to use recapcha.  This library makes sure there is really a user on a webpage making the request, as opposed to some script.  We've used this in the past for perspectiveapi.com

In short - I don't feel this needs to be changed right now, but if you want to make the server a POST I won't object either.

File: capstone/src/main/webapp/youtubeScript.js (hunk 9 above)
Comment (dborkan) [reply capstone/src/main/webapp/youtubeScript.js:25]: Is there a link that documents these category numbers?  If so can you add it to the comments?

File: capstone/src/main/webapp/youtubeScript.js [hunk 20]
```diff
@@ -171,11 +204,83 @@ function showAvailableAttributes() {
...
+      document.getElementById('channelIdForAnalysis').disabled = true;
+    } else {
+      document.getElementById('channelIdForAnalysis').value = "";
+      document.getElementById('channelIdForAnalysis').disabled = false;
+    }
+  }
+}
+
+/** Creates radio buttons to allow the user to select between various categories*/
+function showCategories() {
```
Comment (dborkan) [inline capstone/src/main/webapp/youtubeScript.js:248]: Just FYI:

This is fine for now, but just FYI another way you might be able to achieve the same functionality is by creating all these elements in your .html file under the `categoryContainer` but adding the style `display: none;`.  Then when you want to show the categories you could change it to `display: block` (or similar).  The advantage of that is it might be a little easier to read, although the disadvantage is you wouldn't be creating each category item in a loop in JavaScript.

Also frameworks like Angular/React/etc typically help in cases like this where you may want to create 1 HTML element per item in a JS array.  They let you write HTML templates which get applied to each element in the array.

**Review by dborkan — APPROVED** [review]
Just a few small comments/FYIs but overall looks good

**Comment by olu144** [comment]
I have switched from using a GET request to a POST request to communicate with the YouTube servlets in #40.

**Review by dslucas — APPROVED** [review]

File: capstone/src/main/webapp/youtubeScript.js (hunk 9 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js:25]: Yes, I had added it in "add comment to show YOUTUBE_CATEGORIES source"
